### PR TITLE
Tester mgmt commands

### PR DESCRIPTION
**Pull request by J-coder13**
### Description
Add appdistribution:testers:add and appdistribution:testers:remove commands.

### Inputs
Take a space separated list of emails or a file with a comma separated list of emails.
### Sample Commands
appdistribution:testers:add
--file <file> "a path to a file containing a list of tester emails to be added"

appdistribution:testers:remove
--file <file> "a path to a file containing a list of tester emails to be removed"

### CODE REVIEW

File: src/commands/appdistribution-distribute.ts [hunk 1]
```diff
@@ -73,14 +54,14 @@ module.exports = new Command("appdistribution:distribute <release-binary-file>")
     const appId = getAppId(options.app);
     const distribution = new Distribution(file);
     const releaseNotes = getReleaseNotes(options.releaseNotes, options.releaseNotesFile);
-    const testers = getTestersOrGroups(options.testers, options.testersFile);
-    const groups = getTestersOrGroups(options.groups, options.groupsFile);
-    const requests = new AppDistributionClient(appId);
+    const testers = testerEmailParser(options.testers, options.testersFile);
```
Comment (tagboola) [inline src/commands/appdistribution-distribute.ts:57]: The name of the function `testerEmailParser` should be more generalize since it can be used for more than just parsing tester emails. E.g. on the line below, it's used to get a list of groups.

File: src/commands/appdistribution-distribute.ts [hunk 2]
```diff
@@ -127,7 +108,7 @@ module.exports = new Command("appdistribution:distribute <release-binary-file>")
     utils.logBullet("uploading binary...");
     let releaseName;
     try {
-      const operationName = await requests.uploadRelease(distribution);
+      const operationName = await requests.uploadRelease(appId,distribution);
```
Comment (tagboola) [inline src/commands/appdistribution-distribute.ts:111]: Should be a space before `distribution` i.e. `requests.uploadRelease(appId, distribution)`

File: src/commands/appdistribution-testers-remove.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import * as utils from "../utils";
+import {requireAuth} from "../requireAuth";
+import {FirebaseError} from "../error";
+import {needProjectNumber} from '../projectUtils';
+import {AppDistributionClient} from '../appdistribution/client';
+import {getEmails} from '../testerEmailParser';
+
+
+module.exports = new Command("appdistribution:testers:remove [emails...]")
+    .description("Delete Testers")
```
Comment (tagboola) [inline src/commands/appdistribution-testers-remove.ts:11]: nit: Testers should be lowercase i.e. "Delete testers"

File: src/commands/appdistribution-testers-remove.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    .option(
+        "--file <file>",
+        "a path to a file containing a comma separated list of tester emails to be removed"
+    )
+    .before(requireAuth)
+    .action(async (emails: string[], options?: any) => {
+      const projectNumber = await needProjectNumber(options);
+      const request = new AppDistributionClient();
```
Comment (tagboola) [inline src/commands/appdistribution-testers-remove.ts:19]: Consider changing the variable name `request` to something more relevant.

File: src/commands/appdistribution-testers-remove.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+      const emailsArr = getEmails(emails, options.file);
+      let deleteResponse;
+      try {
+        console.log("Deleting the following testers: " + emailsArr + " from project " + projectNumber);
```
Comment (tagboola) [inline src/commands/appdistribution-testers-remove.ts:23]: `emailsArr` could be a very long array. Consider just printing the count i.e. `Deleting X testers from project Y`

File: src/commands/appdistribution-testers-remove.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        deleteResponse = await request.removeTesters(projectNumber, emailsArr);
+      } catch (err) {
+        throw new FirebaseError(`failed to remove testers.${err}` + {exit: 1});
+      }
+      if (options.debug) {
+        utils.logSuccess(`Testers: ${deleteResponse.emails},have been successfully deleted`);
```
Comment (tagboola) [inline src/commands/appdistribution-testers-remove.ts:29]: I like the idea of printing all the emails in debug mode to give developers more information. However, I don't think you need to do it. I _think_ (please confirm) in debug mode the response for http calls are logged, so developers will already be able to see all the emails in the response

File: src/commands/appdistribution-testers-remove.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+      } else {
+        utils.logSuccess(`${deleteResponse.emails.length} Testers have successfully been deleted`);
```
Comment (tagboola) [inline src/commands/appdistribution-testers-remove.ts:31]: nit: Testers should be lowercase and  i.e. `X testers have successfully been deleted"

File: src/commands/apps-android-sha-create.ts [hunk 8]
```diff
@@ -1,10 +1,10 @@
 import * as clc from "cli-color";
```
Comment (tagboola) [inline src/commands/apps-android-sha-create.ts:1]: Please revert your changes to this file

File: src/operation-poller.ts [hunk 9]
```diff
@@ -10,6 +10,7 @@ export interface OperationPollerOptions {
   backoff?: number;
   maxBackoff?: number;
   masterTimeout?: number;
+  once?: boolean;
```
Comment (tagboola) [inline src/operation-poller.ts:13]: Did you mean to add this? If not please revert the change to this file.

File: src/test/appdistro/client.spec.ts [hunk 10]
```diff
@@ -37,22 +38,59 @@ describe("distribution", () => {
     rimraf.sync(tempdir.name);
   });
 
+  describe("addTesters", () => {
+    let emails = ["a@foo.com", "b@foo.com"];
```
Comment (tagboola) [inline src/test/appdistro/client.spec.ts:42]: nit: Add new line below `emails`

File: src/test/appdistro/client.spec.ts [hunk 11]
```diff
@@ -37,22 +38,59 @@ describe("distribution", () => {
...
+    it('should throw error if upload fails', async () => {
```
Comment (tagboola) [inline src/test/appdistro/client.spec.ts:43]: Please update the name of this test

File: src/test/appdistro/client.spec.ts [hunk 12]
```diff
@@ -37,22 +38,59 @@ describe("distribution", () => {
...
+      nock(api.appDistributionOrigin)
+          .post(`/v1/projects/${projectNumber}/testers:batchAdd`)
+          .reply(200, {});
+      await expect(appDistributionClient.addTesters(projectNumber, emails)).to.be.eventually.fulfilled;
+      expect(nock.isDone()).to.be.true;
+    });
+  })
+
+  describe("deleteTesters", () => {
+    let emails = ["a@foo.com", "b@foo.com"];
```
Comment (tagboola) [inline src/test/appdistro/client.spec.ts:61]: nit: Add new line below `emails`

File: src/testerEmailParser.ts [hunk 13]
```diff
@@ -0,0 +1,31 @@
+import * as fs from 'fs-extra';
```
Comment (tagboola) [inline src/testerEmailParser.ts:1]: Since this is only used by app distribution, consider moving it into an app distribution specific directory instead of `/src`

**Review by tagboola — COMMENTED** [review]
Nice refactoring! Please also fix the Github Action errors.

File: src/appdistribution/client.ts [hunk 14]
```diff
@@ -94,6 +97,7 @@ export class AppDistributionClient {
     return _.get(JSON.parse(apiResponse.body), "name");
   }
 
+
```
Comment (rebehe) [inline src/appdistribution/client.ts:100]: nit: unnecessary newline?

File: src/appdistribution/client.ts [hunk 15]
```diff
@@ -174,4 +178,29 @@ export class AppDistributionClient {
...
+      origin: api.appDistributionOrigin,
+      apiVersion: "v1",
+      auth: true,
+      data,
+    });
+
+    return _.get(apiResponse, "body");
+  }
+
+  async removeTesters(projectNumber: string, emails: string[]) :Promise<BatchRemoveTestersResponse>{
```
Comment (rebehe) [inline src/appdistribution/client.ts:195]: nit: spaces in this method call is weird. It should look like:

```
  async removeTesters(projectNumber: string, emails: string[]): Promise<BatchRemoveTestersResponse> {
```

File: src/commands/appdistribution-distribute.ts (hunk 1 above)
Comment (rebehe) [reply src/commands/appdistribution-distribute.ts:57]: +1. I recommend combining testerEmailParser and ensureFileExists into one helper class, something like `options-parser-util.ts`

File: src/commands/appdistribution-distribute.ts [hunk 16]
```diff
@@ -184,7 +165,7 @@ module.exports = new Command("appdistribution:distribute <release-binary-file>")
         );
       }
     }
-
+
```
Comment (rebehe) [inline src/commands/appdistribution-distribute.ts:168]: nit: unnecessary spaces here?

File: src/commands/appdistribution-testers-add.ts [hunk 17]
```diff
@@ -0,0 +1,28 @@
+import {Command} from "../command";
+import * as utils from "../utils";
+import {requireAuth} from "../requireAuth";
+import {FirebaseError} from "../error";
+import {needProjectNumber} from '../projectUtils';
+import {AppDistributionClient} from '../appdistribution/client';
+import {getEmails} from '../testerEmailParser';
+
+module.exports = new Command("appdistribution:testers:add [emails...]")
+    .description("Create Testers")
```
Comment (rebehe) [inline src/commands/appdistribution-testers-add.ts:10]: Maybe call this "Add testers" so it matches the command

File: src/commands/appdistribution-testers-add.ts [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+    .option(
+        "--file <file>",
+        "a path to a file containing a comma separated list of tester emails to be added"
```
Comment (rebehe) [inline src/commands/appdistribution-testers-add.ts:13]: It's a little weird that we specify comma separated list. It can be comma separated or it can be newline separated. Also weird because the `emails...` property isn't a comma separated list, it's a space separated list.

Basically maybe reword it to be `a path to a file containing a list of tester emails to be added`

File: src/commands/appdistribution-testers-add.ts [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+    )
+    .before(requireAuth)
+    .action(async (emails: string[], options?: any) => {
+      const projectNumber = await needProjectNumber(options);
+      const request = new AppDistributionClient();
+      const emailsArr = getEmails(emails, options.file);
+      try {
+        console.log("Adding the following testers: " + emailsArr + " to project " + projectNumber);
```
Comment (rebehe) [inline src/commands/appdistribution-testers-add.ts:21]: Maybe use the `this is a string variable ${variable}` string interpolation format here instead

File: src/commands/appdistribution-testers-remove.ts (hunk 3 above)
Comment (rebehe) [reply src/commands/appdistribution-testers-remove.ts:11]: Also rename to be "Remove testers" to be more consistent with the command name

File: src/commands/appdistribution-testers-remove.ts [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+import {needProjectNumber} from '../projectUtils';
+import {AppDistributionClient} from '../appdistribution/client';
+import {getEmails} from '../testerEmailParser';
+
+
+module.exports = new Command("appdistribution:testers:remove [emails...]")
+    .description("Delete Testers")
+    .option(
+        "--file <file>",
+        "a path to a file containing a comma separated list of tester emails to be removed"
```
Comment (rebehe) [inline src/commands/appdistribution-testers-remove.ts:14]: Same comment in the other command

File: src/commands/appdistribution-testers-remove.ts (hunk 4 above)
Comment (rebehe) [reply src/commands/appdistribution-testers-remove.ts:19]: +1, and also do the same thing in the other command

File: src/commands/appdistribution-testers-remove.ts (hunk 5 above)
Comment (rebehe) [reply src/commands/appdistribution-testers-remove.ts:23]: +1, and do the same in the other command

File: src/commands/apps-android-sha-create.ts (hunk 8 above)
Comment (rebehe) [reply src/commands/apps-android-sha-create.ts:1]: +1

File: src/operation-poller.ts (hunk 9 above)
Comment (rebehe) [reply src/operation-poller.ts:13]: +1

File: src/testerEmailParser.ts (hunk 13 above)
Comment (rebehe) [reply src/testerEmailParser.ts:1]: +1

File: src/commands/appdistribution-testers-remove.ts (hunk 6 above)
Comment (J-coder13) [reply src/commands/appdistribution-testers-remove.ts:29]: The response is not logged

File: src/testerEmailParser.ts (hunk 13 above)
Comment (J-coder13) [reply src/testerEmailParser.ts:1]: Its used by a couple different utility files.

File: src/appdistribution/options-parser-util.ts [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+  return emails;
+}
+
+export function ensureFileExists(file: string, message = ""): void {
+  if (!fs.existsSync(file)) {
+    throw new FirebaseError(`File ${file} does not exist: ${message}`);
+  }
+}
+
+function testerSplitter(value: string): string[] {
```
Comment (rebehe) [inline src/appdistribution/options-parser-util.ts:31]: Rename to just splitter or something? Since it's also used to split groups

File: src/commands/appdistribution-testers-add.ts [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+  .before(requireAuth)
+  .action(async (emails: string[], options?: any) => {
+    const projectNumber = await needProjectNumber(options);
+    const appDistroClient = new AppDistributionClient();
+    const emailsArr = getEmails(emails, options.file);
+    try {
+      console.log(`Adding ${emailsArr.length} testers to project ${projectNumber}`);
+      await appDistroClient.addTesters(projectNumber, emailsArr);
+    } catch (err) {
+      throw new FirebaseError(`failed to add testers.${err}` + { exit: 1 });
```
Comment (rebehe) [inline src/commands/appdistribution-testers-add.ts:21]: Why is there a period in the string? Maybe there should be a space? Also capitalize f in failed?

File: src/commands/appdistribution-testers-add.ts [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+    }
+    utils.logSuccess(`Testers Created Successfully`);
```
Comment (rebehe) [inline src/commands/appdistribution-testers-add.ts:23]: nit: only capitalize first letter

File: src/commands/appdistribution-testers-add.ts (hunk 22 above)
Comment (rebehe) [reply src/commands/appdistribution-testers-add.ts:21]: Same applies to the other command

File: src/commands/appdistribution-testers-remove.ts [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+    try {
+      console.log(`Deleting ${emailsArr.length} testers from project ${projectNumber}`);
+      deleteResponse = await appDistroClient.removeTesters(projectNumber, emailsArr);
+    } catch (err) {
+      throw new FirebaseError(`failed to remove testers.${err}` + { exit: 1 });
+    }
+    if (options.debug) {
+      utils.logSuccess(`Testers: ${deleteResponse.emails}, have been successfully deleted`);
+    } else {
+      utils.logSuccess(`${deleteResponse.emails.length} testers have successfully been deleted`);
```
Comment (rebehe) [inline src/commands/appdistribution-testers-remove.ts:27]: I believe the emails field isn't set if the response returns no testers. Make sure this is nullsafe

File: src/commands/index.js [hunk 25]
```diff
@@ -14,6 +14,8 @@ module.exports = function (client) {
   };
 
   client.appdistribution = {};
+  client.appdistribution.addTesters = loadCommand("appdistribution-testers-add");
+  client.appdistribution.deleteTesters = loadCommand("appdistribution-testers-remove");
```
Comment (rebehe) [inline src/commands/index.js:18]: To match the other commands, should this be `client.appdistribution.tester.add` and `client.appdistribution.testers.delete`?

Comment (rebehe) [reply src/commands/index.js:18]: Oh wait also it should be `remove`, not `delete`

File: src/appdistribution/client.ts [hunk 26]
```diff
@@ -58,38 +58,44 @@ export interface UploadReleaseResponse {
...
+
 /**
  * Makes RPCs to the App Distribution server backend.
  */
 export class AppDistributionClient {
-  private readonly appName: string;
-
-  constructor(appId: string) {
-    this.appName = `projects/${appId.split(":")[1]}/apps/${appId}`;
+  getAppName(appId: string): string {
```
Comment (tagboola) [inline src/appdistribution/client.ts:69]: Can this be made private?

File: src/appdistribution/options-parser-util.ts (hunk 21 above)
Comment (tagboola) [reply src/appdistribution/options-parser-util.ts]: +1

**Review by tagboola — APPROVED** [review]
Please fix the failing tests and also update `changelog.md` with details on the new commands

Here is a history of the changelog to get an idea of how its updated
https://github.com/firebase/firebase-tools/commits/master/CHANGELOG.md

File: src/test/appdistro/client.spec.ts [hunk 27]
```diff
@@ -37,12 +38,53 @@ describe("distribution", () => {
...
+      await expect(appDistributionClient.addTesters(projectNumber, emails)).to.be.rejected;
+      expect(nock.isDone()).to.be.true;
+    });
+
+    it("should resolve when request succeeds", async () => {
+      nock(api.appDistributionOrigin)
+        .post(`/v1/projects/${projectNumber}/testers:batchAdd`)
+        .reply(200, {});
+      await expect(appDistributionClient.addTesters(projectNumber, emails)).to.be.eventually
+        .fulfilled;
```
[the remark this answers is not in the document]
Comment (J-coder13) [reply src/test/appdistro/client.spec.ts]: This function does not.

File: src/appdistribution/client.ts [hunk 28]
```diff
@@ -174,4 +178,44 @@ export class AppDistributionClient {
...
+      apiVersion: "v1",
+      auth: true,
+    });
+    try {
+      await appDistroV2Client.request({
+        method: "POST",
+        path: `${projectName}/testers:batchAdd`,
+        headers: { "X-Client-Version": `${pkg.name}/${pkg.version}` },
+        body: { emails: emails },
+      });
```
Comment (bkendall) [inline src/appdistribution/client.ts:194]: If you prefer, I _think_ that this can be simplified (and typed) a bit:

```typescript
await appDistroV2Client.post<BatchRemoveTestersResponse, BatchRemoveTestersResponse>(
  `${projectName}/testers:batchAdd`,
  { emails: emails },
  { headers: { "X-Client-Version": `${pkg.name}/${pkg.version}` } });
```

File: src/appdistribution/client.ts [hunk 29]
```diff
@@ -174,4 +178,44 @@ export class AppDistributionClient {
...
+
+  async removeTesters(projectName: string, emails: string[]): Promise<BatchRemoveTestersResponse> {
+    const appDistroV2Client = new Client({
+      urlPrefix: api.appDistributionOrigin,
+      apiVersion: "v1",
+      auth: true,
+    });
+    let apiResponse: ClientResponse<BatchRemoveTestersResponse>;
+    try {
+      apiResponse = await appDistroV2Client.request({
```
Comment (bkendall) [inline src/appdistribution/client.ts:210]: This could also be typed better and use `post`, if you like

File: src/appdistribution/options-parser-util.ts [hunk 30]
```diff
@@ -0,0 +1,56 @@
+import * as fs from "fs-extra";
+import { FirebaseError } from "../error";
+import { needProjectNumber } from "../projectUtils";
+
+/* Takes in comma separated string or a path to a comma/new line separated file
+and converts the input into an string[] of testers or groups. Value takes precedent
+ over file.*/
```
Comment (bkendall) [inline src/appdistribution/options-parser-util.ts:7]: JSDoc-type comments for functions have a pretty well defined pattern:

```suggestion
/**
 * Takes in comma separated string or a path to a comma/new line separated file
 * and converts the input into an string[] of testers or groups. Value takes precedent
 * over file.
 */
```

This makes it so that IDEs can show the docs and use the information in it. Definitely apply this to the one below too.

File: src/commands/appdistribution-testers-add.ts [hunk 31]
```diff
@@ -0,0 +1,23 @@
+import { Command } from "../command";
+import * as utils from "../utils";
+import { requireAuth } from "../requireAuth";
+import { FirebaseError } from "../error";
+import { AppDistributionClient } from "../appdistribution/client";
+import { getEmails, getProjectName } from "../appdistribution/options-parser-util";
+import { needProjectNumber } from "../projectUtils";
+
+module.exports = new Command("appdistribution:testers:add [emails...]")
+  .description("Add testers to project")
```
Comment (bkendall) [inline src/commands/appdistribution-testers-add.ts:10]:
```suggestion
  .description("add testers to project")
```

File: src/commands/appdistribution-testers-remove.ts [hunk 32]
```diff
@@ -0,0 +1,33 @@
+import { Command } from "../command";
+import * as utils from "../utils";
+import { requireAuth } from "../requireAuth";
+import { FirebaseError } from "../error";
+import { needProjectNumber } from "../projectUtils";
+import { AppDistributionClient } from "../appdistribution/client";
+import { getEmails } from "../appdistribution/options-parser-util";
+
+module.exports = new Command("appdistribution:testers:remove [emails...]")
+  .description("Remove testers")
```
[the remark this answers is not in the document]
Comment (bkendall) [reply src/commands/appdistribution-testers-remove.ts:11]:
```suggestion
  .description("removes testers from a project")
```

**Review by bkendall — CHANGES_REQUESTED** [review]
Couple small things, but LGTM otherwise

File: src/appdistribution/client.ts (hunk 28 above)
Comment (J-coder13) [reply src/appdistribution/client.ts:194]: I was having a hard time getting the right response value from it afterwards so I think that the way we have it currently is fine.

Comment (bkendall) [reply src/appdistribution/client.ts:194]: Let me know what the difficulty was at some point (maybe message me directly) and let's see if we can't improve it some way. :)

**Review by bkendall — APPROVED** [review]

**Comment by bkendall** [comment]
Please be sure to "Squash and merge" when you're ready!
